IPv6 Inner/Outer Hardware Offload

evelineraine opened the pull request.

swx-jenkins3 commented:
Can one of the admins verify this patch?

yuskul commented:
bot:retest

yuskul commented:
bot:retest

yuskul commented:
bot:retest

swx-jenkins3 commented:
Test FAILed.

swx-jenkins3 commented:
Test PASSed.

swx-jenkins3 commented:
Test PASSed.

swx-jenkins3 commented:
Test FAILed.

swx-jenkins3 commented:
Test FAILed.

swx-jenkins3 commented:
Test PASSed.

swx-jenkins3 commented:
Tests PASSed.

swx-jenkins3 commented:
Tests PASSed.

swx-jenkins3 commented:
Tests PASSed.

swx-jenkins3 commented:
Tests FAILed.

swx-jenkins3 commented:
Tests PASSed.

swx-jenkins3 commented:
Tests FAILed for https://github.com/Mellanox/OVS/pull/26.
See http://hpc-master.lab.mtl.com:8080/job/_OVS_team/job/OVS_Baidu/149//console for details.

swx-jenkins3 commented:
Tests PASSed for PR#https://github.com/Mellanox/OVS/pull/26.
See http://hpc-master.lab.mtl.com:8080/job/_OVS_team/job/OVS_Baidu/150/console for details.

asafpenso commented on lib/netdev-rte-offloads.c:
```diff
@@ -1289,6 +1287,29 @@ add_dpdk_flow_patterns(struct flow_patterns *patterns,
 
+    /* IP v6 */
+    if (match->flow.dl_type == htons(ETH_TYPE_IPV6)) {
+        memset(&specs->ipv6, 0, sizeof(specs->ipv6));
```
For now, it's good. Once you prepare the patches for the community need to remove the () from sizeof.
E.g: instead of sizeof(specs->ipv6) write sizeof specs->ipv6.

evelineraine replied:
*Fixed in the new patches*

asafpenso commented on lib/netdev-rte-offloads.c:
```diff
@@ -1289,6 +1287,29 @@ add_dpdk_flow_patterns(struct flow_patterns *patterns,
+
+        masks->ipv6.hdr.proto = match->wc.masks.nw_proto;
+        masks->ipv6.hdr.hop_limits = match->wc.masks.nw_ttl;
+        rte_memcpy(masks->ipv6.hdr.src_addr, match->wc.masks.ipv6_src.s6_addr, sizeof(masks->ipv6.hdr.src_addr));
```
Have you run checkpatch? looks like a long line that exceeds 80 chars.

evelineraine replied:
*Fixed in the new patches*

asafpenso commented on lib/netdev-rte-offloads.c:
```diff
@@ -1762,7 +1783,29 @@ add_vport_vxlan_flow_patterns(struct flow_patterns *patterns,
+        proto = specs->ipv6.hdr.proto &
+                masks->ipv6.hdr.proto;
     } else {
+        VLOG_ERR("tunnel L3 header is neither IPv4 nor IPv6");
```
If there was not a print before, we should keep it without it. In case of an error the terminal would be spammed with messages and reduced the perf. 
So, either use VLOG_ERR_RL which Roni added to limit the rate, or remove the print.

evelineraine replied:
In the netdev_dpdk_validate_flow() we are already checking which offload patterns we do and don't support. After this check, all invalid patterns for *outer* header are filtered out. This leaves us with either IPv4 or IPv6 as outer L3 protocols. If for some reason here we get neither, this is a real error.
`VLOG_ERR_RL` would be an option. Any reason we shouldn't make this a default and use everywhere in place of `VLOG_ERR`?

evelineraine replied:
Changed to VLOG_ERR_RL in new commit

asafpenso commented on lib/netdev-rte-offloads.c:
```diff
@@ -1401,8 +1400,11 @@ netdev_dpdk_add_rte_flow_offload(struct netdev_rte_port *rte_port,
     int result = -1;
     struct netdev_rte_port * vport = NULL;
 
+    VLOG_DBG("Adding rte offload for flow ufid "UUID_FMT, UUID_ARGS((struct uuid *) ufid));
```
I think all prints start with lower case letter, please confirm

evelineraine replied:
Some start with an uppercase letter. For example see `lib/dpif-netdev.c#2351`. Is there a convention we follow?

evelineraine replied:
I also checked `lib/netdev-dpdk.c`. There, all log messages start with an uppercase letter. Since our `lib/netdev-rte-offloads.c` branched from it, we probably should follow the uppercase convention for consistency.

asafpenso commented on lib/netdev-rte-offloads.c:
```diff
@@ -1738,7 +1738,7 @@ add_vport_vxlan_flow_patterns(struct flow_patterns *patterns,
     /* IP v4 */
     uint8_t proto = 0;
-    if (match->flow.dl_type == htons(ETH_TYPE_IP)) {
+    if (match->wc.masks.tunnel.ip_src || match->wc.masks.tunnel.ip_dst) {
```
Can you explain which issue you resolved here and why existing code is not good?

evelineraine replied:
This code runs on already decapped packet. When looking at flow.dl_type you are looking at the inner L3 protocol. This code block needs to determine the outer L3 protocol.
Since tunnel struct does not have tunnel L3 type, we look at which masks are set (for IPv4 src/dst or for IPv6 src/dst (in the following patches)).

evelineraine commented:
> Please see my comments.
> In general, make sure to run checkpatch and update the code accordingly.

Yes, I actually forgot to do that.

> Besides the commit header, need also to add text to the body, giving more info on the patch itself.
> For now, it's ok and once you prepare the patches for the community need to take that into consideration.
> The commit header should start with a capital letter after the ':' and should end with '.'.

Ok, though in general the first line of commit message shouldn't end with period, and there isn't any note about this in OVS contributing guidelines. But I see _most_ commits end with a period, so it's a convention I didn't know about, I'll be following it.

I don't suppose there is a way to amend a commit inside a PR and not lose reviewers' comments?

asafpenso commented:
With our newly code we indeed use VLOG_ERR_RL.

Regards,
Asaf Penso

From: Eveline Raine <notifications@github.com>
Sent: Sunday, March 24, 2019 12:59 PM
To: Mellanox/OVS <OVS@noreply.github.com>
Cc: Asaf Penso <asafp@mellanox.com>; Review requested <review_requested@noreply.github.com>
Subject: Re: [Mellanox/OVS] IPv6 Inner/Outer Hardware Offload (#26)


@evelineraine commented on this pull request.

________________________________

In lib/netdev-rte-offloads.c<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMellanox%2FOVS%2Fpull%2F26%23discussion_r268428296&data=02%7C01%7Casafp%40mellanox.com%7C30a16ee7bbdf49c2032808d6b047a79a%7Ca652971c7d2e4d9ba6a4d149256f461b%7C0%7C0%7C636890219136555657&sdata=2jm%2BGUaicpAiHbQqmR6WtHMlimdeyowN0HRmdi%2F9B0o%3D&reserved=0>:

>      } else {

+        VLOG_ERR("tunnel L3 header is neither IPv4 nor IPv6");

In the netdev_dpdk_validate_flow() we are already checking which offload patterns we do and don't support. After this check, all invalid patterns for outer header are filtered out. This leaves us with either IPv4 or IPv6 as outer L3 protocols. If for some reason here we get neither, this is a real error.
VLOG_ERR_RL would be an option. Any reason we shouldn't make this a default and use everywhere in place of VLOG_ERR?

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMellanox%2FOVS%2Fpull%2F26%23discussion_r268428296&data=02%7C01%7Casafp%40mellanox.com%7C30a16ee7bbdf49c2032808d6b047a79a%7Ca652971c7d2e4d9ba6a4d149256f461b%7C0%7C0%7C636890219136565622&sdata=%2BsgMmiFWSRNat1zItEUTPG2nikn2zVSFilNTTErRiAQ%3D&reserved=0>, or mute the thread<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAoqtvFSDaptk3aAOwsrphRLYFnQ9Lwfnks5vZ1pWgaJpZM4b4YWo&data=02%7C01%7Casafp%40mellanox.com%7C30a16ee7bbdf49c2032808d6b047a79a%7Ca652971c7d2e4d9ba6a4d149256f461b%7C0%7C0%7C636890219136565622&sdata=qk3ZXdyIwLUsRLh%2Fxr3O7ZHfxlWi%2B6MHY%2F4EGQX%2FP7U%3D&reserved=0>.

asafpenso commented:
Ophir and I discussed that in the past and we prefer to update the patches themselves, and not adding a new patch to fix the comments.
Afterwards to push -f , which will maintain the comments for the previous commits and will show the new patches as well.
I hope it is clear 😊

Regards,
Asaf Penso

From: Eveline Raine <notifications@github.com>
Sent: Sunday, March 24, 2019 1:12 PM
To: Mellanox/OVS <OVS@noreply.github.com>
Cc: Asaf Penso <asafp@mellanox.com>; Review requested <review_requested@noreply.github.com>
Subject: Re: [Mellanox/OVS] IPv6 Inner/Outer Hardware Offload (#26)


Please see my comments.
In general, make sure to run checkpatch and update the code accordingly.

Yes, I actually forgot to do that.

Besides the commit header, need also to add text to the body, giving more info on the patch itself.
For now, it's ok and once you prepare the patches for the community need to take that into consideration.
The commit header should start with a capital letter after the ':' and should end with '.'.

Ok, though in general the first line of commit message shouldn't end with period, and there isn't any note about this in OVS contributing guidelines. But I see most commits end with a period, so it's a convention I didn't know about, I'll be following it.

I don't suppose there is a way to amend a commit inside a PR and not lose reviewers' comments?

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMellanox%2FOVS%2Fpull%2F26%23issuecomment-475948834&data=02%7C01%7Casafp%40mellanox.com%7C1adf6787bfdf45b9f17608d6b0497eab%7Ca652971c7d2e4d9ba6a4d149256f461b%7C0%7C0%7C636890227042735054&sdata=EN3hxX2HTK5YZBRVS3mCRhqxHIAVwjTlr281hkP6rQg%3D&reserved=0>, or mute the thread<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAoqtvMQFxKgSLEjS3ehQGlvtc3k0PpZUks5vZ11tgaJpZM4b4YWo&data=02%7C01%7Casafp%40mellanox.com%7C1adf6787bfdf45b9f17608d6b0497eab%7Ca652971c7d2e4d9ba6a4d149256f461b%7C0%7C0%7C636890227042745059&sdata=S1K6N%2Fc%2FVVwEBeSITDMsdtY3rRZh6uKbL9M73apro3A%3D&reserved=0>.

OphirMunk commented:
Comments regarding commit: ("dpif-netdev: Improve ufid logging")

1. Same comment as before (and for the remaining commits): commit message body should be filled in.
A title only is not sufficient.

2. A comment added inside the code 

3. Please fix checkpatch warnings:


== Checking cc409080a341 ("dpif-netdev: Improve ufid logging") ==
WARNING: Line is 81 characters long (recommended limit is 79)
#20 FILE: lib/dpif-netdev.c:2155:
VLOG_WARN("Masked megaflow ufid "UUID_FMT" is not associated with a mark?\n",

WARNING: Line is 117 characters long (recommended limit is 79)
#38 FILE: lib/dpif-netdev.c:2187:
VLOG_DBG("Associated dp_netdev flow ufid "UUID_FMT" with mark %u", UUID_ARGS((struct uuid *) &flow->ufid), mark);

Lines checked: 45, Warnings: 2, Errors: 0

OphirMunk commented on lib/dpif-netdev.c:
```diff
@@ -2184,7 +2184,7 @@ mark_to_flow_associate(const uint32_t mark, struct dp_netdev_flow *flow)
     flow->mark = mark;
 
-    VLOG_DBG("Associated dp_netdev flow %p with mark %u\n", flow, mark);
+    VLOG_DBG("Associated dp_netdev flow ufid "UUID_FMT" with mark %u", UUID_ARGS((struct uuid *) &flow->ufid), mark);
```
Why did you drop end of line "\n"? If by mistake - need to restore.

evelineraine replied:
Actually VLOG_* already adds `\n` for messages automatically if not present. I will restore it, but it does not matter if we add `\n` or not.

evelineraine replied:
*Fixed in the new patches*

OphirMunk commented on lib/netdev-rte-offloads.c:
```diff
@@ -898,7 +898,7 @@ dump_flow_pattern(struct rte_flow_item *item)
         VLOG_DBG("rte flow ipv4 pattern:\n");
         if (ipv4_spec) {
-            VLOG_DBG("  Spec: tos=0x%"PRIx8", ttl=%"PRIx8", proto=0x%"PRIx8
+            VLOG_DBG("  Spec: tos=0x%"PRIx8", ttl=0x%"PRIx8", proto=0x%"PRIx8
```
In OVS ttl is printed as unsigned char. So the fix should be:
```
 VLOG_DBG("  Spec: tos=0x%"PRIx8", ttl=%"PRIu8", proto=0x%"PRIx8
```

evelineraine replied:
*Fixed in the new patches*

OphirMunk commented on lib/netdev-rte-offloads.c:
```diff
@@ -909,7 +909,7 @@ dump_flow_pattern(struct rte_flow_item *item)
         }
         if (ipv4_mask) {
-            VLOG_DBG("  Mask: tos=0x%"PRIx8", ttl=%"PRIx8", proto=0x%"PRIx8
+            VLOG_DBG("  Mask: tos=0x%"PRIx8", ttl=0x%"PRIx8", proto=0x%"PRIx8
```
Same comment here regarding ttl

evelineraine replied:
*Fixed in the new patches*

OphirMunk commented on lib/netdev-rte-offloads.c:
```diff
@@ -1401,8 +1400,11 @@ netdev_dpdk_add_rte_flow_offload(struct netdev_rte_port *rte_port,
     int result = -1;
     struct netdev_rte_port * vport = NULL;
 
+    VLOG_DBG("Adding rte offload for flow ufid "UUID_FMT, UUID_ARGS((struct uuid *) ufid));
```
When adding VLOG messages on an expected positive case we may be overwhelmed with printouts. For example during high insertion rate. Instead of VLOG printouts we should take care of adding statistics that will inform of the successful operation.

evelineraine replied:
Without this printout, I discovered, it is difficult to determine the location of errors during offloading. Besides, we are printing multiple debug messages with all the flow patterns we add. How does adding this message make it worse?

OphirMunk commented on lib/netdev-rte-offloads.c:
```diff
@@ -1812,6 +1814,9 @@ netdev_vport_vxlan_add_rte_flow_offload(struct netdev_rte_port * rte_port,
 {
+    VLOG_DBG("Adding rte offload for vport vxlan flow ufid "UUID_FMT,
+        UUID_ARGS((struct uuid *) ufid));
+
```
Similar comment here regarding adding VLOG printouts on expected positive cases

evelineraine replied:
*(Same as above)*

swx-jenkins3 commented:
Tests PASSed for PR#26.
See http://hpc-master.lab.mtl.com:8080/job/_OVS_team/job/OVS_Baidu/169/console for details.

swx-jenkins3 commented:
Tests PASSed for PR#26.
See http://hpc-master.lab.mtl.com:8080/job/_OVS_team/job/OVS_Baidu/170/console for details.

evelineraine commented:
> Regarding commit: ("packets: Add IPv6 printf pattern")
> 
> 1. Commit title: printf --> printout
> 2. Should elaborate in the commit message body about the contents of
>    the commit. A title only is not sufficient.
> 3. This is a preceding commit to another one which is several commits
>    away: "netdev-rte-offloads: Add flow dumping for IPv6".
>    The two commits should be consecutive.
> 4. Please note in file lib/packets.c that there is already an API for
>    printing IPv6 addresses:
> 
> ```
> /* Stores the string representation of the IPv6 address 'addr' into the
>  * character array 'addr_str', which must be at least INET6_ADDRSTRLEN
>  * bytes long. If addr is IPv4-mapped, store an IPv4 dotted-decimal string. */
>  const char *
>  ipv6_string_mapped(char *addr_str, const struct in6_addr *addr)
> ```
> Let's keep the same format in our code. Can you use this API for Baidu
> code?

1-3: Fixed 
#### 4: 
I know this function exists. However there is another approach - to use `*_FMT` macros. I think this has a better precedent - we use this approach to print UUIDs in log messages like so:
```c
VLOG_WARN("Masked megaflow ufid "UUID_FMT
                  " is not associated with a mark?\n",
                  UUID_ARGS((struct uuid *)mega_ufid));
```

In addition to this we already have such formatting macros in `lib/packets.h` for IPv4:
```c
#define IP_FMT "%"PRIu32".%"PRIu32".%"PRIu32".%"PRIu32
#define IP_ARGS(ip)                             \
    ntohl(ip) >> 24,                            \
    (ntohl(ip) >> 16) & 0xff,                   \
    (ntohl(ip) >> 8) & 0xff,                    \
    ntohl(ip) & 0xff
```

So this makes sense to me to add parity for IPv6.

However the two approaches differ in that `*_FMT` macros always prints 39 characters (zeroes or not) but `ipv6_string_mapped()` can compress addresses.

> Comments regarding commit: ("dpif-netdev: Improve ufid logging")
> 
> 1. Same comment as before (and for the remaining commits): commit message body should be filled in.
>    A title only is not sufficient.
> 2. A comment added inside the code
> 3. Please fix checkpatch warnings:
> 
> == Checking cc409080a341 ("dpif-netdev: Improve ufid logging") ==
> WARNING: Line is 81 characters long (recommended limit is 79)
> #20 FILE: lib/dpif-netdev.c:2155:
> VLOG_WARN("Masked megaflow ufid "UUID_FMT" is not associated with a mark?\n",
> 
> WARNING: Line is 117 characters long (recommended limit is 79)
> #38 FILE: lib/dpif-netdev.c:2187:
> VLOG_DBG("Associated dp_netdev flow ufid "UUID_FMT" with mark %u", UUID_ARGS((struct uuid *) &flow->ufid), mark);
> 
> Lines checked: 45, Warnings: 2, Errors: 0

1,3: Fixed

evelineraine commented:
bot:retest

swx-jenkins3 commented:
Tests PASSed for PR#26.
See http://hpc-master.lab.mtl.com:8080/job/_OVS_team/job/OVS_Baidu/173/console for details.

OphirMunk commented:
> > ```
> > /* Stores the string representation of the IPv6 address 'addr' into the
> >  * character array 'addr_str', which must be at least INET6_ADDRSTRLEN
> >  * bytes long. If addr is IPv4-mapped, store an IPv4 dotted-decimal string. */
> >  const char *
> >  ipv6_string_mapped(char *addr_str, const struct in6_addr *addr)
> > ```
> > Let's keep the same format in our code. Can you use this API for Baidu
> > code?
> 
> 
> 
> #### 4:
> 
> I know this function exists. However there is another approach - to use `*_FMT` macros. I think this has a better precedent - we use this approach to print UUIDs in log messages like so:
> 
> 
> ```c
> VLOG_WARN("Masked megaflow ufid "UUID_FMT
>                   " is not associated with a mark?\n",
>                   UUID_ARGS((struct uuid *)mega_ufid));
> ```

There is a difference between ufid and IPv6 addresses. The former is condense. Every bit counts. We want to see all bits. The later is very sparse. We want to see only the few non-zero bits. This is why a standard was defined for printing IPv6 addresses in RFC 4291. https://tools.ietf.org/html/rfc4291. 
The API used in OVS for printing IPv6 addresses is 
```
ipv6_string_mapped()
```
and it is used dozens of times and uniquely in the OVS code. Can you please use it?
We should follow the OVS community standards. This is one of them.

> In addition to this we already have such formatting macros in `lib/packets.h` for IPv4:
> 
> ```c
> #define IP_FMT "%"PRIu32".%"PRIu32".%"PRIu32".%"PRIu32
> #define IP_ARGS(ip)                             \
>     ntohl(ip) >> 24,                            \
>     (ntohl(ip) >> 16) & 0xff,                   \
>     (ntohl(ip) >> 8) & 0xff,                    \
>     ntohl(ip) & 0xff
> ```

The end result is that IPv4 is printed according to the IPv4 standard even though it uses this formatting macros. We are not sparse for IPv4. Still for IPv6 we need to be sparse.

> However the two approaches differ in that `*_FMT` macros always prints 39 characters (zeroes or not) but `ipv6_string_mapped()` can compress addresses.

Yes...in order to be sparse.

OphirMunk commented:
> Changes the print format from `PRIx8` to `PRIu8`.
>

Whenever there is a fix for a previous commit, just before Signed-off-by ... at the end of the commit message you should add:
```
Fixes: <12 characters hash> ("<commit title>"). 
```
For example:
```
Fixes: b10d46a60013 ("tests: Check dpif-netdev odp_actions consistency.")
Signed-off-by: Ian Stokes <ian.stokes@intel.com>
```
You can find the fixed commit details by running 'git blame "file_name"', then receive the fixed commit SHA1 number, then run 'git show "SHA1 number"' for the full details

OphirMunk commented:
> Previously the code looked at decapsulated packet's Ethernet type to
> determine tunnel L3 protocol is IPv4. The fix involves checking tunnel
> address masks for IPv4 source and destination. If non-zero, the protocol
> is IPv4.

Same comment as before: there is a fix here, so should add "Fixes: ..." at the end of commit message just before "Signed-of-by..." line.

OphirMunk commented:
> OphirMunk 2 days ago  Member
> When adding VLOG messages on an expected positive case we may be overwhelmed with printouts. 
> For example during high insertion rate. Instead of VLOG printouts we should take care of adding
> statistics that will inform of the successful operation.
> 
> evelineraine 8 hours ago  Author Member
> Without this printout, I discovered, it is difficult to determine the location of errors during offloading. 

Have you checked an option to enhance error messages during offloading such that you it will be easy to determine the location of errors during offloading?

> Besides, we are printing multiple debug messages with all the flow patterns we add. How does adding 
> this message make it worse?

One of our goals is to be best in flow insertion rate. VLOG_DBG adds additional code (even with a low log level and no printouts) and may reduce the insertion rate (should be measured and verified). If you are right and there are multiple VLOG_DBG on the fast path - they should all be removed as well (but don't do it yourself now). Any VLOG_DBG that is on  an error path or on a non-fast path - is OK. We are only measured on successful fast path. 
By the way, looking at netdev-dpdk.c before it was split into netdev-rte-offloads.c - I notice no VLOG_DBG on the fast path.

OphirMunk commented:
> packets: Add IPv6 printout pattern.
>
As mentioned previously we should use
```
ipv6_string_mapped()
```
for printing IPv6 addresses. Therefore I suggest omitting this commit.

swx-jenkins3 commented:
Tests PASSed for PR#26.
See http://hpc-master.lab.mtl.com:8080/job/_OVS_team/job/OVS_Baidu/176/console for details.

evelineraine commented:
@OphirMunk 
I have changed flow dumping to print compresses IPv6 addresses using `ipv6_string_mapped()` and the removed the commit that adds `*_FMT` patterns.
I have also added the `Fixes` tags to my commits. Thank you for pointing that out!

I still don't agree about debug prints. In our code (in addition to multiline pattern dumps) we have this message printed for each flow being offloaded:
```c
VLOG_DBG("second flow of decap created");
```
Without it, we wouldn't be able to debug all the issues we did on virtio since February. We have this line, same as the lines I added, to increase visibility into what is actually happening inside OVS. 
It does impact performance, of course. In the same way counting marked and hardware offloaded packets that you added does as well. But it was done for the same reason - to increase visibility.

Regarding the fast path vs slow path - flows offloading is not really fastpath, just like inserting rules into the kernel module in OVS-kernel is not fastpath either. I don't think it makes sense to optimize it as hard as to drop debug information in order to save cycles. In `dpif-netdev.c` several debug messages are being printed for each flow even before it reaches `netdev-rte-offloads.c`. 
If we want to squeeze the last bit of performance (while sacrificing some visibility in live production environments), we should add a way to compile-out debug messages with a `#define` flag like DPDK does.

All this being said, we really want to have this PR merged ASAP. If you still think I should remove the prints, I will and then we can follow up on this discussion later.

OphirMunk commented:
Merged into hwol-mlnx-dpdk17.11 branch
